### ISSUE TITLE
DB crashed after routine operation due to guarantee failed

### ISSUE THREAD

**Issue by qusiax** [rethinkdb/rethinkdb#4611]
2015-08-01T06:32:48.436125013 462060.834928s error: Error in src/extproc/js_runner.cc at line 178:
    2015-08-01T06:32:48.450451318 462060.848512s error: Guarantee failed: [fn_id != __null] 
    2015-08-01T06:32:48.450466521 462060.848527s error: Backtrace:
    2015-08-01T06:32:48.993639350 462061.391702s error: Sat Aug  1 06:32:48 2015\n\n1: backtrace_t::backtrace_t() at ??:?\n2: format_backtrace(bool) at ??:?\n3: report_fatal_error(char const*, int, char const*, ...) at ??:?\n4: js_runner_t::call(std::string const&, std::vector<ql::datum_t, std::allocator<ql::datum_t> > const&, js_runner_t::req_config_t const&) at ??:?\n5: ql::js_func_t::call(ql::env_t*, std::vector<ql::datum_t, std::allocator<ql::datum_t> > const&, ql::eval_flags_t) const at ??:?\n6: ql::func_t::call(ql::env_t*, ql::datum_t, ql::eval_flags_t) const at ??:?\n7: ql::map_trans_t::lst_transform(ql::env_t*, std::vector<ql::datum_t, std::allocator<ql::datum_t> >*, ql::datum_t const&) at ??:?\n8: ql::ungrouped_op_t::operator()(ql::env_t*, std::map<ql::datum_t, std::vector<ql::datum_t, std::allocator<ql::datum_t> >, optional_datum_less_t, std::allocator<std::pair<ql::datum_t const, std::vector<ql::datum_t, std::allocator<ql::datum_t> > > > >*, ql::datum_t const&) at ??:?\n9: rget_cb_t::handle_pair(scoped_key_value_t&&, concurrent_traversal_fifo_enforcer_signal_t) at ??:?\n10: concurrent_traversal_adapter_t::handle_pair_coro(scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t) at ??:?\n11: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (concurrent_traversal_adapter_t::*)(scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t)> (concurrent_traversal_adapter_t*, scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t)> >::run_action() at ??:?\n12: coro_t::run() at ??:?
    2015-08-01T06:32:48.993715075 462061.391775s error: Exiting.

This happened after 4 days of uptime with all commands having been tested. the error log doesn't look to have enough information for me to debug on the application level. Any tips?

**Comment by danielmewes** [comment]
Hi @qusiax, thank you for reporting this.
From the looks of it, this is related to the execution of an `r.js` statement. It definitely looks like a server bug.

Here's the reformatted version of the backtrace:
```
Error in src/extproc/js_runner.cc at line 178:
Guarantee failed: [fn_id != __null] 
Backtrace:
Sat Aug  1 06:32:48 2015

1: backtrace_t::backtrace_t() at ??:?
2: format_backtrace(bool) at ??:?
3: report_fatal_error(char const*, int, char const*, ...) at ??:?
4: js_runner_t::call(std::string const&, std::vector<ql::datum_t, std::allocator<ql::datum_t> > const&, js_runner_t::req_config_t const&) at ??:?
5: ql::js_func_t::call(ql::env_t*, std::vector<ql::datum_t, std::allocator<ql::datum_t> > const&, ql::eval_flags_t) const at ??:?
6: ql::func_t::call(ql::env_t*, ql::datum_t, ql::eval_flags_t) const at ??:?
7: ql::map_trans_t::lst_transform(ql::env_t*, std::vector<ql::datum_t, std::allocator<ql::datum_t> >*, ql::datum_t const&) at ??:?
8: ql::ungrouped_op_t::operator()(ql::env_t*, std::map<ql::datum_t, std::vector<ql::datum_t, std::allocator<ql::datum_t> >, optional_datum_less_t, std::allocator<std::pair<ql::datum_t const, std::vector<ql::datum_t, std::allocator<ql::datum_t> > > > >*, ql::datum_t const&) at ??:?
9: rget_cb_t::handle_pair(scoped_key_value_t&&, concurrent_traversal_fifo_enforcer_signal_t) at ??:?
10: concurrent_traversal_adapter_t::handle_pair_coro(scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t) at ??:?
11: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (concurrent_traversal_adapter_t::*)(scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t)> (concurrent_traversal_adapter_t*, scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t)> >::run_action() at ??:?
12: coro_t::run() at ??:?
```

@Tryneus any ideas about this one?

**Comment by Tryneus** [comment]
It appears there is an unhandled case where we timeout when compiling a function - it's expecting an id to be returned, but it's getting an error instead.  Instead of guaranteeing, we should pass the error back up.

**Comment by qusiax** [comment]
Would that be beneficial to a user? Even if I knew that this exception
happened is there a user side action I could take at runtime?

It's worth noting that this happened on a non mutating query that's run
basically on the hot path of the code (so it was a run a bunch of times
before failing).

Is there a way to instruct the compiler to cache compilations? Or a way to
precompile statically rather than passing in a js string?

**Comment by Tryneus** [comment]
@qusiax, I can't tell at the moment why compiling the function timed out in this case, but we do cache the compiled functions already.  The cache may be cleared if there is a problem communicating with the worker process, but you should see errors in the server's log if and when that happens.

It is worth noting that the cache is not shared between queries, so the cache is only an advantage if the `r.js` function is used multiple times in the same query.

**Comment by qusiax** [comment]
I certainly don't have the experience with the code to know whether this is feasible or not, but could js compilations be cached across queries? 

This would probably introduce new concerns around memory usage required to maintain an LRU cache. Alternatively, rethinkdb could upon request return a binary blob which can be provided back by the user upon a subsequent query, moving the memory management requirements to the user. I imagine official clients could introduce a convenient client side caching layer.

**Comment by danielmewes** [comment]
Moving this to 2.1 for scheduling purposes.

**Comment by Tryneus** [comment]
Fix for the crash in review 3129 - this situation will return an error instead.

**Comment by Tryneus** [comment]
The fix has been approved and merged to `next` in commit a437aed, and cherry-picked into v2.1.x and v2.0.x in commits 589bc21 and 3845ec4, respectively.
